// kelonio/out/etc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STATE_FILE = ".kelonio.state.json";
var HEADER_SIDE = "- ".repeat(17).trim();
exports.HEADER = HEADER_SIDE + " Performance " + HEADER_SIDE;
exports.FOOTER = "- ".repeat(40).trim();
var fs;
var canUseFs = true;
try {
    fs = require("fs");
}
catch (_a) {
    canUseFs = false;
}
var BenchmarkFileState = /** @class */ (function () {
    function BenchmarkFileState() {
        if (!canUseFs) {
            throw new Error("Unable to access file system");
        }
    }
    BenchmarkFileState.prototype.exists = function () {
        return fs.existsSync(exports.STATE_FILE);
    };
    BenchmarkFileState.prototype.read = function () {
        return JSON.parse(fs.readFileSync(exports.STATE_FILE, "utf-8"));
    };
    BenchmarkFileState.prototype.write = function (data) {
        fs.writeFileSync(exports.STATE_FILE, JSON.stringify(data), "utf-8");
    };
    BenchmarkFileState.prototype.delete = function () {
        try {
            fs.unlinkSync(exports.STATE_FILE);
        }
        catch (_a) { }
    };
    return BenchmarkFileState;
}());
exports.BenchmarkFileState = BenchmarkFileState;
